maas/resource_maas_boot_source_selection.go

```diff
@@ -0,0 +1,227 @@
+		Release:   d.Get("release").(string),
+		Arches:    convertToStringSlice(d.Get("arches").(*schema.Set).List()),
+		Subarches: convertToStringSlice(d.Get("subarches").(*schema.Set).List()),
+		Labels:    convertToStringSlice(d.Get("labels").(*schema.Set).List()),
+	}
+
+	bootsourceselection, err := client.BootSourceSelections.Create(bootsource.ID, &bootsourceselectionParams)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId((fmt.Sprintf("%v", bootsourceselection.ID)))
+
+	return resourceBootSourceSelectionRead(ctx, d, meta)
+}
+
+func resourceBootSourceSelectionRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ClientConfig).Client
+
+	bootsource, err := getBootSource(client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	bootsourceselection, err := getBootSourceSelection(client, bootsource.ID, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	tfState := map[string]interface{}{
+		"arches":         bootsourceselection.Arches,
+		"boot_source_id": bootsource.ID,
+		"labels":         bootsourceselection.Labels,
+		"os":             bootsourceselection.OS,
+		"release":        bootsourceselection.Release,
+		"subarches":      bootsourceselection.Subarches,
+	}
+	if err := setTerraformState(d, tfState); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+func resourceBootSourceSelectionUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ClientConfig).Client
+
+	bootsource, err := getBootSource(client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	bootsourceselection, err := getBootSourceSelection(client, bootsource.ID, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	bootsourceselectionParams := entity.BootSourceSelectionParams{
+		OS:        d.Get("os").(string),
+		Release:   d.Get("release").(string),
+		Arches:    convertToStringSlice(d.Get("arches").(*schema.Set).List()),
+		Subarches: convertToStringSlice(d.Get("subarches").(*schema.Set).List()),
+		Labels:    convertToStringSlice(d.Get("labels").(*schema.Set).List()),
+	}
+
+	if _, err := client.BootSourceSelection.Update(bootsource.ID, bootsourceselection.ID, &bootsourceselectionParams); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceBootSourceSelectionRead(ctx, d, meta)
+}
+
+func resourceBootSourceSelectionDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ClientConfig).Client
+
+	bootsource, err := getBootSource(client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	bootsourceselection, err := getBootSourceSelection(client, bootsource.ID, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
```

Comment:
Here if we are not the default OS, we simply delete. If we are the default OS we are updating.